### PR TITLE
DOC:`Building the NumPy API and reference docs` rewrite

### LINKED ISSUE #19104: DOC: `howto_build_docs` needs rewrite
As per the [documentation team meeting](https://hackmd.io/oB_boakvRqKR-_2jRV-Qjg) on 24th May, [Building the NumPy API and reference docs](https://numpy.org/devdocs/docs/howto_build_docs.html) needs to be updated with the following: 

1. Include build guidelines for Windows users; recommend Gitpod and WSL app
2. Explanation for building NumPy before the docs 
3. Remove/modify the submodules section
4. Clarify use of `make latex` and `make dist`

I think points 1 and 2 can be addressed with a dedicated prerequisites section like in [Building from source](https://numpy.org/devdocs/user/building.html#prerequisites).

From the meeting, I gathered that there are no submodules for the docs currently but may re-appear in the future because of mathjax (please correct me if I am wrong). In that case should I remove/comment out the section for the time being? 

I also understood that `make latex` and `make dist` are not usually required for developing the documentation. Should I clarify that they are mostly for release purposes?

cc: @melissawm @rossbar

### PR DESCRIPTION
**Pull request by Mukulikaa**
1. Made distinct sections for prerequisites and instructions
2. Added alternatives for Windows users
3. Included explanation for building NumPy before docs

Closes #19104

### CODE REVIEW

File: doc/source/docs/howto_build_docs.rst
Comment (melissawm) [inline doc/source/docs/howto_build_docs.rst:50]: Actually, in the `doc_requirements.txt` file, we are asking for `sphinx==4.0.1`, but (surprisingly) in `environment.yml` we are using `sphinx=3.5.4`. I'm assuming anything above 3.5.4 works - @rossbar can you help us here?

File: doc/source/docs/howto_build_docs.rst
Comment (melissawm) [inline doc/source/docs/howto_build_docs.rst:19]: Maybe we can qualify this as "not natively available on Windows" or similar. There are a number of solutions to using make on windows and they might work, we just don't want to recommend any in particular.

File: doc/source/docs/howto_build_docs.rst
Comment (melissawm) [inline doc/source/docs/howto_build_docs.rst:36]:
```suggestion
   Note that you can e.g. install NumPy to a temporary location and set
```

File: doc/source/docs/howto_build_docs.rst
Comment (melissawm) [inline doc/source/docs/howto_build_docs.rst:52]: There are other places - for example in the SVD tutorial SciPy is also needed. Maybe just say "some examples and documents require the use of SciPy functions".

**Review by melissawm — COMMENTED** [review]
Hi @Mukulikaa ! Thanks for doing this - I have some minor comments but in general this looks good to me! If any newcomers have suggestions for instructions they'd like to see on this page, they are encouraged to contribute their ideas, too!

File: doc/source/docs/howto_build_docs.rst
Comment (mattip) [reply doc/source/docs/howto_build_docs.rst:50]: I didn't update the `environment.yml` when I updated `doc_requirements.txt`, and we have no check that the requirements are in sync. It would be nice to have an automatic way to update the docs, maybe we could `include` the relevant lines of the `doc_requirements.txt` or alternatively suggest users do `pip install -r doc_requirements.txt` without mentioning exact versions?

Comment (Mukulikaa) [reply doc/source/docs/howto_build_docs.rst:50]: I think not specifying the exact versions will be better for the maintenance of the doc. Since users don't have to separately install `Sphinx`, I don't think it will cause any problems.

**Comment by Mukulikaa** [comment]
Thanks for the review! I've made the required changes.

**Review by melissawm — APPROVED** [review]
Looks good to me, let's wait for @rossbar 's review :)

**Comment by Mukulikaa** [comment]
I'll try to resolve the merge conflict by incorporating Mars' commit [efaed19](https://github.com/numpy/numpy/pull/19185/commits/efaed1950223b5906dd751b04104b4b9191a929b) before Ross' review.

**Review by rossbar — COMMENTED** [review]
Overall LGTM, though I'd prefer not to advertise WSL so strongly --- my reasoning here is that GitPod has in some sense been "officially" adopted by NumPy, and we offer some semblance of support for developing in that environment; the same can't be said for WSL.

It's not a blocker, but I personally think it'd be an improvement to drop the `Building Locally` subsection and only mention WSL in passing.

**Comment by Mukulikaa** [comment]
Makes sense. I will remove the section and just mention it's an option for persistent local set-up.

**Review by rossbar — COMMENTED** [review]
I took the opportunity to replace a couple external links with intersphinx links as well - overall this is looking good!

File: doc/source/docs/howto_build_docs.rst
Comment (Mukulikaa) [reply doc/source/docs/howto_build_docs.rst:42]: That is a good idea! I'll make the changes.

File: doc/source/docs/howto_build_docs.rst
Comment (Mukulikaa) [reply doc/source/docs/howto_build_docs.rst:82]: I had asked about this in the issue #19104 and @melissawm had directed me to the mathjax issue #19036. From the discussion there I figured that the submodule may come back in the future and so we have to keep the section in the doc? I must confess I didn't follow the entire discussion so I was not sure whether to keep the section or not.

Comment (rossbar) [reply doc/source/docs/howto_build_docs.rst:82]: The lines above this are correct --- NumPy does use a submodule again; but these lines are specific to older submodules that are no longer used.

Comment (Mukulikaa) [reply doc/source/docs/howto_build_docs.rst:82]: Thanks for the clarification!

Comment (rossbar) [reply doc/source/docs/howto_build_docs.rst]: The second sentence (about getting numpydoc on pypi) is also no longer relevant and should be removed - numpy itself now gets numpydoc from pypi rather than including it as a submodule. It looks like you can eliminate the PyPI and numpydoc links as well.

Comment (Mukulikaa) [reply doc/source/docs/howto_build_docs.rst]: Will it be worth it to add a line explaining that the submodule is for `mathjax`, which helps us render mathematical notation? We could also add a line in the dependencies section saying that we use the `numpydoc` Sphinx extension to render the doctrings.

Comment (rossbar) [reply doc/source/docs/howto_build_docs.rst]: > Will it be worth it to add a line explaining that the submodule is for mathjax

I would just as soon avoid this, so we avoid this whole problem again in the future if we decide to remove the mathjax submodule :)

> We could also add a line in the dependencies section saying that we use the numpydoc Sphinx extension to render the doctrings.

:+1:

File: doc/source/docs/howto_build_docs.rst
Comment (rossbar) [reply doc/source/docs/howto_build_docs.rst]: > I'm assuming anything above 3.5.4 works

AFAIK there's no reason to pin to 3.5.4 - I've been building with ~=4 for a while now without problems

> and we have no check that the requirements are in sync.

I'm not sure what the best practices are for keeping `environments.yml` and `requirements.txt` in sync. I think there's a way in `environments.yml` files to specify to use `pip` for certain dependencies... maybe that supports requirements files? I'm not really sure...

**Review by rossbar — APPROVED** [review]
LGTM, thanks @Mukulikaa !

I will merge once the CI is done.

**Pull request merged**
